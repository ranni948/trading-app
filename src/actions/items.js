import uuid from 'uuid';
import database, { storage } from '../firebase/firebase';

//ADD_EXPENSE
export const addItem = (item) => ({
    type: 'ADD_ITEM',
    item
});

export const startAddItem = (itemData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            description = '', 
            price = 0, 
            category = '', 
            location = '',
            images = []
        } = itemData;
        const item = { title, description, price, category, location, images };

        return database.ref(`users/${uid}/items`).push(item).then((ref) => {      

            dispatch(addItem({
                id: ref.key,
                userId: uid,
                ...item
            }));
        });
    };
};

//REMOVE_ITEM
export const removeItem = ({ id } = {}) => ({
    type: 'REMOVE_ITEM',
    id
});

export const startRemoveItem = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        database.ref(`users/${uid}/items/${id}`).once('value').then((snapshot) => {
            return snapshot.val();
        }).then((result) => {
            //delete item from DB first so that any reference to the images is gone, if something
            //goes wrong with deletinf the images it won;t matter since there is no reference to them anymore
            database.ref(`users/${uid}/items/${id}`).remove().then(() => {
                dispatch(removeItem({ id }));
            }).then(() => {
                result.images.map((image, index) => {
                    storage.ref(`images/${image.filename}`).delete();
                });
            });
    
        }); 
    }
}

//EDIT_ITEM
export const editItem = (id, updates) => ({
    type: 'EDIT_ITEM',
    id,
    updates
});

export const startEditItem = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/items/${id}`).update(updates).then(() => {
            dispatch(editItem(id, updates));
        });
    }
}

// SET_ITEMS
export const setItems = (items) => ({
    type: 'SET_ITEMS',
    items
});

export const startSetItems = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users`).once('value').then((snapshot) => {
            const items = [];

            snapshot.forEach((child) => {
                child.forEach((child1) => {
                    child1.forEach((childSnapshot) => {
                        items.push({
                            id: childSnapshot.key,
                            userId: child.key,
                            ...childSnapshot.val()
                        });
                    })
                })
            });
            dispatch(setItems(items));
        });
    };
};