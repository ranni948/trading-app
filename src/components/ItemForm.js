import React from 'react';
//import moment from 'moment';
import FileUploader from 'react-firebase-file-uploader';
import { storage } from '../firebase/firebase';

export default class ItemForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.item ? props.item.title : '',
            description: props.item ? props.item.description : '',
            price: props.item ? (props.item.price / 100).toString() : '',
            category: props.item ? (props.item.category) : '',
            location: props.item ? (props.item.location) : '',
            error: '',
            filename: '',
            isUploading: false,
            progress: 0,
            images: props.item ? (props.item.images) : []
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }))
    };
    onPriceChange = (e) => {
        const price = e.target.value;
        if (!price || price.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ price }))
        }
    };
    onCategoryChange = (e) => {
        const category = e.target.value;
        this.setState(() => ({ category }))
    };
    onLocationChange = (e) => {
        const location = e.target.value;
        this.setState(() => ({ location }))
    };
    handleUploadStart = () => this.setState({isUploading: true, progress: 0});
    handleProgress = (progress) => this.setState({progress});
    handleUploadError = (error) => {
        this.setState({isUploading: false});
        console.error(error);
    }
    handleUploadSuccess = (filename) => {
        this.setState({filename, progress: 100, isUploading: false});
        // this.setState((prevState) => ({
        //     images: [...prevState.images, filename] 
        // }));
        storage.ref('images').child(filename).getDownloadURL().then(url => this.setState((prevState) => ({ 
            images: [...prevState.images, {
                url
            }] 
        })));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.price || !this.state.category || !this.state.location) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                description: this.state.description,
                price: parseFloat(this.state.price, 10) * 100,
                category: this.state.category,
                location: this.state.location,
                images: this.state.images
            })
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {
                    this.state.isUploading && <p>Progress: {this.state.progress}</p>
                }
                {
                    this.state.images.map((image, index) => (
                        <img src={image.url} key={index}/>
                    ))
                }
                <FileUploader
                    accept="image/*"
                    name="item"
                    randomizeFilename
                    multiple
                    storageRef={storage.ref('images')}
                    onUploadStart={this.handleUploadStart}
                    onUploadError={this.handleUploadError}
                    onUploadSuccess={this.handleUploadSuccess}
                    onProgress={this.handleProgress}
                />
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="What are you trading?"
                    autoFocus
                    className="text-input"
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <input
                    type="text"
                    placeholder="Price"
                    className="text-input"
                    value={this.state.price}
                    onChange={this.onPriceChange}
                />
                <input
                    type="text"
                    placeholder="Select Category"
                    className="text-input"
                    value={this.state.category}
                    onChange={this.onCategoryChange}
                />
                <input
                    type="text"
                    placeholder="Location"
                    className="text-input"
                    value={this.state.location}
                    onChange={this.onLocationChange}
                />
                <input
                    type="text"
                    placeholder="Decription"
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <div>
                    <button className="button">Save Item</button>
                </div>
            </form>
        )
    }
}