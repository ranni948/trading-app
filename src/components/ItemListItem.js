import React from 'react';
import { NavLink } from 'react-router-dom';
import numeral from 'numeral';

const ItemListItem = ({item}) => (
    <NavLink className="list-items list-item" to={`/view/${item.id}`}>
        <img className="list-item-image" src={item.images[0].url}/>
        <span className="list-item__price">{numeral(item.price / 100).format('$0,0.00')}</span>
        <div>{item.title}</div>
    </NavLink>
);

export default ItemListItem;