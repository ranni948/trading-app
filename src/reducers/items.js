//Items Reducer
const itemsReducerDefaultState = []

export default (state = itemsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_ITEM':
        console.log('Action items', action.item);
        console.log('State', state);
            return [
                ...state,
                action.item
            ]
        case 'SET_ITEMS':
            return action.items
        // case 'REMOVE_EXPENSE':
        //     return state.filter(({ id }) => id !== action.id)
        // case 'EDIT_EXPENSE':
        //     return state.map((expense) => {
        //         if (expense.id === action.id) {
        //             return {
        //                 ...expense,
        //                 ...action.updates
        //             }
        //         } else {
        //             return expense;
        //         };
        //     });
        default:
            return state
    }
};
