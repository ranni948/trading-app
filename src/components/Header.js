import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header =(props) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>BarterBro</h1>
                </Link>
                <div className="header__content">
                    <img className="image-profile"src={props.userPhoto} />
                    <span className="button button--link">{
                        props.userName
                        }</span>
                    <button className="button button--link" onClick={props.startLogout}>SIGN-OUT</button>
                </div>
            </div>
        </div>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
    return {
        userPhoto: state.auth.userPhoto,
        userName: state.auth.userName
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);