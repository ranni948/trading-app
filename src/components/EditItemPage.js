import React from 'react';
import { connect } from 'react-redux';
import ItemForm from './ItemForm';
import { startEditItem } from '../actions/items';
import { storage } from '../firebase/firebase';

export class EditItemPage extends React.Component {
    onSubmit = (item) => {
        item.images.map(image => {
            if (image.deleted) {
                console.log("Deleted " + image.filename);
                storage.ref(`images/${image.filename}`).delete().then(() => {
                    //Do Nothing
                }).catch((error) => {
                    console.error(error);
                    return;
                });
            }
        })
        item.images = item.images.filter((image) => !image.deleted);
        item.images.map(image => {
            console.log("Deleted " + image.filename);
        });
        this.props.startEditItem(this.props.item.id, item);
        this.props.history.push('/');
    };
    render() {
        return (
            <div className="content">
                <div className="page-header">
                    <div className="content-container">
                    <h1 className="page-header__title">Edit Item</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ItemForm 
                        item={this.props.item}
                        operation={true}
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startEditItem: (id, item) => dispatch(startEditItem(id, item))
});

const mapStateToProps = (state, props) => ({
    item: state.items.find((item) => item.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditItemPage);