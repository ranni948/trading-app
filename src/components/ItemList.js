import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import ItemListItem from './ItemListItem';

export const ItemList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <Link className="button" to="/create">+ Add Item</Link>
        </div>
        <div className="list-body">
        <div className="list-item--message">
                        <span>Items</span>
                    </div>
        {
            props.items.length === 0 ? (
                <div className="list-items list-item--message">
                    <span>No items</span>
                </div>
            ) : (
                <div className="list-items">
                    {console.log(props.items)}
                    {
                        props.items.map((item) => {
                        return <ItemListItem key={item.id} item={item} />
                    })}
                </div>
            )
        }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return  {
        items: state.items
    };
};

export default connect(mapStateToProps)(ItemList);