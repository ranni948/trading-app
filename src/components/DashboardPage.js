import React from 'react';
import ItemList from './ItemList';

const DashboardPage = () => (
    <div className="content">
        <ItemList />
    </div>
);

export default DashboardPage;