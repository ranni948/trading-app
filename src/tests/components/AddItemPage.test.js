import React from 'react';
import { shallow } from 'enzyme';
import { AddItemPage } from '../../components/AddItemPage';
import items from '../fixtures/items'

test('should render add item page correctly', () => {
    const wrapper = shallow(<AddItemPage startAddItem={() => { }}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should call startAddItem on button click', () => {
    const startAddItemSpy = jest.fn();
    let history = { push: jest.fn() }
    const wrapper = shallow(<AddItemPage startAddItem={startAddItemSpy} history={history}/>);
    wrapper.find('ItemForm').prop('onSubmit')(items[1]);
    expect(startAddItemSpy).toHaveBeenLastCalledWith(items[1]);
});