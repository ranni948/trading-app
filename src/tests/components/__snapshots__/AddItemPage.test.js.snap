// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add item page correctly 1`] = `
<div
  className="content"
>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Item
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ItemForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
